File: VAR.R
# ===========================
# 1. Cargar librerías
# ===========================
install.packages(c("vars", "readr", "dplyr"))
library(vars)
library(readr)
library(dplyr)

# ===========================
# 2. Cargar CSV
# ===========================
# Cambia la ruta si no está en tu working directory
ruta <- "VAR_Financieras.csv"

# Detectar separador y decimales
# Si tus números tienen coma de miles y punto decimal:
df <- read_delim(ruta, delim = ",", locale = locale(decimal_mark = "."))

# Revisar primeras filas
head(df)
str(df)

# ===========================
# 3. Convertir columnas a numéricas
# ===========================
# Eliminar comas de miles
cols <- colnames(df)[-1]  # asumimos que la primera columna es Fecha
for (col in cols) {
  df[[col]] <- as.numeric(gsub(",", "", df[[col]]))
}

# Eliminar columnas vacías
df <- df %>% select(where(~!all(is.na(.))))

# ===========================
# 4. Convertir precios a rendimientos
# ===========================
# Detectar precios (todos positivos y cambios grandes)
max_diff <- max(diff(as.matrix(df[cols])), na.rm = TRUE)

if(all(df[cols] > 0) & max_diff > 1) {
  cat("Se detectaron PRECIOS. Calculando rendimientos...\n")
  df_ret <- df
  for(col in cols) {
    df_ret[[col]] <- c(NA, diff(log(df[[col]])))  # rendimientos logarítmicos
  }
  df_ret <- na.omit(df_ret)
} else {
  cat("Se detectaron RENDIMIENTOS. Usando datos directamente...\n")
  df_ret <- df %>% select(all_of(cols))
}

# Revisar dimensiones
cat("Filas y columnas después de limpieza:", dim(df_ret), "\n")

# ===========================
# 5. Ajustar VAR
# ===========================
# Convertir a ts (opcional)
df_ts <- ts(df_ret, frequency = 1)

# Ajustar VAR con 2 rezagos
var_model <- VAR(df_ts, p = 2, type = "const")
summary(var_model)

# ===========================
# 6. Funciones Impulso-Respuesta (IRF)
# ===========================
irf_model <- irf(var_model, impulse = cols, response = cols, n.ahead = 10, boot = TRUE)
print(irf_model)

# Graficar IRF
plot(irf_model)

File: VAR.py
import pandas as pd
import matplotlib.pyplot as plt
from statsmodels.tsa.api import VAR

# ===========================
# 1. Subir archivo (Colab) o usar local
# ===========================
try:
    from google.colab import files
    uploaded = files.upload()
    filename = list(uploaded.keys())[0]
except:
    filename = "VAR_Financieras.csv"  # Cambia si no estás en Colab

# ===========================
# 2. Leer CSV
# ===========================
df = pd.read_csv(filename, parse_dates=True, index_col=0)

print("=== Primeras filas del archivo ===")
print(df.head())

# ===========================
# 3. Limpiar columnas: eliminar comas de miles y convertir a float
# ===========================
for col in df.columns:
    df[col] = df[col].astype(str).str.replace(',', '', regex=False)
    df[col] = pd.to_numeric(df[col], errors='coerce')

# Eliminar columnas vacías
df = df.dropna(axis=1, how='all')

# Verificar que haya suficientes columnas numéricas
numeric_cols = df.select_dtypes(include='number').columns
if len(numeric_cols) < 2:
    raise ValueError("Se necesitan al menos 2 columnas numéricas para un VAR.")
df = df[numeric_cols]

# ===========================
# 4. Limpiar filas con NaN
# ===========================
df = df.dropna()
if df.empty:
    raise ValueError("El DataFrame quedó vacío después de limpiar NaN.")

print("\nFilas y columnas después de limpieza:", df.shape)

# ===========================
# 5. Detectar precios o rendimientos
# ===========================
if (df > 0).all().all() and (df.diff().abs().max().max() > 1):
    if df.shape[0] < 2:
        raise ValueError("No hay suficientes filas para calcular rendimientos.")
    print("\nSe detectaron PRECIOS. Calculando rendimientos...")
    df = (df / df.shift(1) - 1).dropna()
else:
    print("\nSe detectaron RENDIMIENTOS. Usando datos directamente.")

if df.empty:
    raise ValueError("No hay datos suficientes después de preparar rendimientos.")

print("\n=== Primeras filas de la base lista para VAR ===")
print(df.head())

# ===========================
# 6. Ajustar modelo VAR
# ===========================
model = VAR(df)
results = model.fit(maxlags=2)
print("\n=== Resumen del VAR ===")
print(results.summary())

# ===========================
# 7. Funciones Impulso-Respuesta (IRF)
# ===========================
irf = results.irf(10)  # horizonte de 10 pasos
print("\n=== Primeros valores de IRF (primeros 5 pasos) ===")
print(irf.irfs[:5])

# Graficar IRF
irf.plot(orth=False)
plt.suptitle("Funciones Impulso-Respuesta (IRF)", fontsize=14)
plt.show()

